File: api/admin/menu/v1/menu.go
package v1

import "github.com/gogf/gf/v2/frame/g"

type MenuTreeReq struct {
	g.Meta `path:"/menu/tree" method:"get" tags:"菜单" summary:"菜单树" permission:"menu:tree"`
}

type MenuTreeRes struct {
	List []*MenuTreeNode `json:"list" dc:"菜单树列表"`
}

type MenuTreeNode struct {
	Id            int64           `json:"id" dc:"菜单ID"`
	Name          string          `json:"name" dc:"菜单名称"`
	Permission    string          `json:"permission" dc:"权限标识"`
	Type          int             `json:"type" dc:"菜单类型：1目录 2菜单 3按钮"`
	Sort          int             `json:"sort" dc:"排序值"`
	ParentId      int64           `json:"parent_id" dc:"父级ID"`
	Path          string          `json:"path" dc:"路由路径"`
	Icon          string          `json:"icon" dc:"图标"`
	Component     string          `json:"component" dc:"组件路径"`
	ComponentName string          `json:"component_name" dc:"组件名称"`
	Status        int             `json:"status" dc:"状态：0正常 1停用"`
	Visible       bool            `json:"visible" dc:"是否可见"`
	KeepAlive     bool            `json:"keep_alive" dc:"是否缓存"`
	AlwaysShow    bool            `json:"always_show" dc:"是否总是显示"`
	Children      []*MenuTreeNode `json:"children,omitempty" dc:"子菜单"`
}

type MenuCreateReq struct {
	g.Meta        `path:"/menu" method:"post" tags:"菜单" summary:"新增菜单" permission:"menu:create"`
	Name          string `json:"name" v:"required#err.invalid_params" dc:"菜单名称"`
	Permission    string `json:"permission" dc:"权限标识"`
	Type          int    `json:"type" dc:"菜单类型：1目录 2菜单 3按钮"`
	Sort          int    `json:"sort" dc:"排序值"`
	ParentId      int64  `json:"parent_id" dc:"父级ID"`
	Path          string `json:"path" dc:"路由路径"`
	Icon          string `json:"icon" dc:"图标"`
	Component     string `json:"component" dc:"组件路径"`
	ComponentName string `json:"component_name" dc:"组件名称"`
	Status        int    `json:"status" dc:"状态：0正常 1停用"`
	Visible       bool   `json:"visible" dc:"是否可见"`
	KeepAlive     bool   `json:"keep_alive" dc:"是否缓存"`
	AlwaysShow    bool   `json:"always_show" dc:"是否总是显示"`
}

type MenuCreateRes struct {
	Id int64 `json:"id" dc:"菜单ID"`
}

type MenuUpdateReq struct {
	g.Meta        `path:"/menu/{id}" method:"put" tags:"菜单" summary:"更新菜单" permission:"menu:update"`
	Id            int64   `json:"id" in:"path" v:"required#err.invalid_params" dc:"菜单ID"`
	Name          *string `json:"name" dc:"菜单名称"`
	Permission    *string `json:"permission" dc:"权限标识"`
	Type          *int    `json:"type" dc:"菜单类型：1目录 2菜单 3按钮"`
	Sort          *int    `json:"sort" dc:"排序值"`
	ParentId      *int64  `json:"parent_id" dc:"父级ID"`
	Path          *string `json:"path" dc:"路由路径"`
	Icon          *string `json:"icon" dc:"图标"`
	Component     *string `json:"component" dc:"组件路径"`
	ComponentName *string `json:"component_name" dc:"组件名称"`
	Status        *int    `json:"status" dc:"状态：0正常 1停用"`
	Visible       *bool   `json:"visible" dc:"是否可见"`
	KeepAlive     *bool   `json:"keep_alive" dc:"是否缓存"`
	AlwaysShow    *bool   `json:"always_show" dc:"是否总是显示"`
}

type MenuUpdateRes struct{}

type MenuDeleteReq struct {
	g.Meta `path:"/menu/{id}" method:"delete" tags:"菜单" summary:"删除菜单" permission:"menu:delete"`
	Id     int64 `json:"id" in:"path" v:"required#err.invalid_params" dc:"菜单ID"`
}

type MenuDeleteRes struct{}

File: manage/src/api/menu.ts
import request from './request'

/** 与后端 system_menu.type 一致：1 目录 2 菜单 3 按钮（侧栏不展示） */
export const MENU_TYPE_DIR = 1
export const MENU_TYPE_MENU = 2
export const MENU_TYPE_BUTTON = 3

export interface MenuTreeNode {
  id: number
  name: string
  permission: string
  type: number
  sort: number
  parent_id: number
  path: string
  icon: string
  component: string
  component_name: string
  status: number
  visible: boolean
  keep_alive?: boolean
  always_show?: boolean
  children?: MenuTreeNode[]
}

/** 侧栏用：去掉按钮节点，并去掉过滤后无子级的目录 */
export function filterSidebarMenus(nodes: MenuTreeNode[]): MenuTreeNode[] {
  const mapped = nodes
    .filter((n) => n.type !== MENU_TYPE_BUTTON)
    .map((n) => {
      const children = n.children?.length ? filterSidebarMenus(n.children) : undefined
      return { ...n, children }
    })
  return mapped.filter((n) => n.type !== MENU_TYPE_DIR || !!n.path || (n.children?.length ?? 0) > 0)
}

export function getUserMenus() {
  return request.get<unknown, { data?: { list?: MenuTreeNode[] } }>('/admin/me/menus')
}

/** 管理端完整菜单树（含按钮），用于角色授权等 */
export function fetchAdminMenuTree() {
  return request.get<unknown, { data?: { list?: MenuTreeNode[] } }>('/admin/menu/tree')
}

export function createAdminMenu(payload: Record<string, unknown>) {
  return request.post<unknown, { data?: { id?: number } }>('/admin/menu', payload)
}

export function updateAdminMenu(id: number, payload: Record<string, unknown>) {
  return request.put<unknown, unknown>(`/admin/menu/${id}`, payload)
}

export function deleteAdminMenu(id: number) {
  return request.delete<unknown, unknown>(`/admin/menu/${id}`)
}
